src/control/system/faultdomain.go

```diff
@@ -267,6 +277,31 @@ func (t *FaultDomainTree) mergeTree(toBeMerged *FaultDomainTree) {
 	return
 }
 
+// RemoveDomain removes a given fault domain from the tree.
+func (t *FaultDomainTree) RemoveDomain(domain *FaultDomain) error {
+	if t == nil {
+		return errors.New("nil FaultDomainTree")
+	}
+	if domain == nil {
+		return nil // nothing to remove
+	}
+	if domain.Equals(t.Domain) {
+		return errors.New("cannot remove root fault domain from tree")
+	}
+	for i := len(t.Children) - 1; i >= 0; i-- {
+		child := t.Children[i]
+		if child.Domain.Equals(domain) {
+			// found it
+			t.Children = append(t.Children[:i], t.Children[i+1:]...)
+			return nil
+		}
+		if child.Domain.IsAncestorOf(domain) {
+			_ = child.RemoveDomain(domain)
```

Comment:
Is it worth checking the error here just in case a future revision changes the logic such that removing a child's domain could fail?

Reply:
Good point -- we should probably return this back up the stack if it does fail somehow at some lower level.

Reply:
Actually, as I think about it, we should return here anyway. If we find the branch we're looking for, we won't find the domain on any other branch of the tree.